src/green.rs

```diff
@@ -2,65 +2,170 @@ use std::{mem::size_of, sync::Arc};
 
 use crate::{SmolStr, TextUnit, Types};
 
-/// `GreenNode` is an immutable syntax tree,
-/// which is cheap to update. It lacks parent
-/// pointers and information about offsets.
+/// Internal node in the immutable tree.
 #[derive(Debug)]
-pub struct GreenNode<T: Types>(GreenNodeImpl<T>);
+pub struct GreenNode<T: Types> {
+    kind: T::Kind,
+    text_len: TextUnit,
+    children: Arc<[GreenElement<T>]>,
+}
+
+impl<T: Types> GreenNode<T> {
+    /// Creates new Node.
+    pub fn new(kind: T::Kind, children: Box<[GreenElement<T>]>) -> GreenNode<T> {
+        let text_len = children.iter().map(|x| x.text_len()).sum::<TextUnit>();
+        GreenNode { kind, text_len, children: children.into() }
+    }
+
+    /// Kind of this node.
+    pub fn kind(&self) -> T::Kind {
+        self.kind
+    }
+
+    /// Length of the text, covered by this node.
+    pub fn text_len(&self) -> TextUnit {
+        self.text_len
+    }
+    /// Children of this node.
+    pub fn children(&self) -> &[GreenElement<T>] {
+        &self.children
+    }
+
+    /// Number of memory bytes of occupied by subtree rooted at `self`.
+    pub(crate) fn memory_size_of_subtree(&self) -> usize {
+        let mut res = size_of::<Self>();
+        self.children().iter().for_each(|el| match el {
+            GreenElement::Token(token) => {
+                res += size_of::<GreenToken<T>>();
+                if token.text.len() > 22 {
```

Comment:
22?

Reply:
Yep, that”s the length when we start to heap-allocat small strings

Reply:
Can that be thrown into a constant somewhere?

Reply:
Or a function in SmolStr, `is_inline()`

Reply:
good call!